### PR TITLE
add pagination & hotfix intermediate graph

### PR DESCRIPTION
**Pull request by colorjam**

### CODE REVIEW

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 1]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
 
     componentDidMount(): void {
         window.addEventListener('resize', this.onWindowResize);
+        this.setState(() => this.updateData())
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:570]: why updateData function need to wrtie in setState's callback?

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 2]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+        const perPageOptions = [
+            { key: '10', text: '10 items per page'},
+            { key: '20', text: '20 items per page'},
+            { key: '50', text: '50 items per page'},
+            { key: 'all', text: 'all items'},
+        ];
+        const horizontalGapStackTokens: IStackTokens = {
+            childrenGap: 20,
+            padding: 10,
+          };
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:636]: It's just a const variable, pls remove this from `render()`

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 3]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
             isShowCustomizedModal, copyTrialId, intermediateOption, sortMessage
         } = this.state;
         const { columnList } = this.props;
-        const tableSource: Array<TableRecord> = JSON.parse(JSON.stringify(this.state.tableSourceForSort));
+        const tableSource = this.state.tableSourceForSort
+        const perPageOptions = [
+            { key: '10', text: '10 items per page'},
+            { key: '20', text: '20 items per page'},
+            { key: '50', text: '50 items per page'},
+            { key: 'all', text: 'all items'},
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:631]: All items

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 4]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+        // eslint-disable-next-line no-debugger
+        // debugger
+
+        this.setState({
+            tableSourceForSort: tableSlice,
+            pageCount: Math.ceil(tableSource.length / this.state.perPage),
+        });
+    }
+    
+    handlePageClick = (e): void => {
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:597]: add parameter e type?

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 5]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+        const tableSource: Array<TableRecord> = JSON.parse(JSON.stringify(this.props.tableSource));
+
+        const tableSlice = tableSource.slice(this.state.offset, this.state.offset + this.state.perPage)
+        // eslint-disable-next-line no-debugger
+        // debugger
+
+        this.setState({
+            tableSourceForSort: tableSlice,
+            pageCount: Math.ceil(tableSource.length / this.state.perPage),
+        });
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:594]: const { xx , xx , xxx } = this.state;

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 6]
```diff
@@ -599,7 +660,35 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+                        breakLabel={"..."}
+                        breakClassName={"break-me"}
+                        pageCount={this.state.pageCount}
+                        marginPagesDisplayed={2}
+                        pageRangeDisplayed={2}
+                        onPageChange={this.handlePageClick}
+                        containerClassName={(this.props.tableSource.length == 0 ? "pagination hidden" : "pagination" )}
+                        subContainerClassName={"pages pagination"}
+                        disableInitialCallback={false}
+                        activeClassName={"active"}/>
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:685]: string type could directly write like this breakClassName="break-me"? Such as other className, label type. pls check

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 7]
```diff
@@ -19,10 +19,12 @@ import { contentStyles, iconButtonStyles } from '../Buttons/ModalTheme';
 import '../../static/style/search.scss';
 import '../../static/style/tableStatus.css';
 import '../../static/style/logPath.scss';
-import '../../static/style/search.scss';
 import '../../static/style/table.scss';
 import '../../static/style/button.scss';
 import '../../static/style/openRow.scss';
+import '../../static/style/pagination.css';
+import ReactPaginate from 'react-paginate';
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:26]: Could you remove this line to line 8?

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 8]
```diff
@@ -96,15 +103,21 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+            pageCount: 0,
+            tableSourceForSort: JSON.parse(JSON.stringify(this.props.tableSource)),
         };
     }
 
     // sort for table column
     onColumnClick = (ev: React.MouseEvent<HTMLElement>, getColumn: IColumn): void => {
         const { tableColumns } = this.state;
-        const { tableSource } = this.props;
+        // const { tableSource } = this.props;
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:119]: delete or put dataForsort here

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 9]
```diff
@@ -554,27 +567,74 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
+            });
+        }
+    }
+
+    updateData(): void {
+        const tableSource: Array<TableRecord> = JSON.parse(JSON.stringify(this.props.tableSource));
+
+        const tableSlice = tableSource.slice(this.state.offset, this.state.offset + this.state.perPage)
+        // eslint-disable-next-line no-debugger
+        // debugger
```
Comment (Lijiaoa) [inline src/webui/src/components/trial-detail/TableList.tsx:589]: It's useful?

File: src/webui/src/static/style/pagination.css [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  border-color: #0071bc;
+  outline: none;
+}
+.pagination > li > a, .pagination > li > span{
+  color: #0071bc
+}
+
+.pagination > li:first-child > a, .pagination > li:first-child > span, .pagination > li:last-child > a, .pagination > li:last-child > span{
+  border-radius: unset
+}
```
Comment (Lijiaoa) [inline src/webui/src/static/style/pagination.css:36]: use scss instead of css

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 11]
```diff
@@ -599,7 +659,35 @@ class TableList extends React.Component<TableListProps, TableListState> {
                         selectionMode={SelectionMode.multiple}
                         selection={this.getSelectedRows}
                     />
+                {
```
Comment (ultmaster) [inline src/webui/src/components/trial-detail/TableList.tsx:662]: I think this is redundant?

File: src/webui/src/components/trial-detail/TableList.tsx [hunk 12]
```diff
@@ -96,15 +109,19 @@ class TableList extends React.Component<TableListProps, TableListState> {
...
             allColumnList: this.getAllColumnKeys(),
-            tableSourceForSort: this.props.tableSource,
-            sortMessage: { field: '', isDescend: false }
+            sortMessage: { field: '', isDescend: false },
+            offset: 0,
+            data: [],
+            perPage: 20,
+            currentPage: 0,
+            pageCount: 0,
+            tableSourceForSort: JSON.parse(JSON.stringify(this.props.tableSource)),
```
Comment (ultmaster) [inline src/webui/src/components/trial-detail/TableList.tsx:118]: Why stringify and parse?

File: src/webui/src/static/style/pagination.scss [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            cursor: pointer;
+            color: #0071bc
+        }
+        span {
+            color: #0071bc
+        }
+    }
+}
+
+// .pagination > e
```
Comment (ultmaster) [inline src/webui/src/static/style/pagination.scss:38]: Please delete the unused part.

**Pull request merged**